File: backend_node/src/controllers/artista.controller.ts
import { Request, Response } from 'express'
import { uploadS3Base64 } from '../configs/s3.config';
import { execute_procedure } from '../configs/mysql.config';


export const insertarArtista = async(req: Request, res: Response) => {
    const { nombre, fecha, nombreFoto, foto} = req.body;
    let ulrFoto = '-'; 
  
    try {
        //guardo al artista en la base de datos
        const insert_result = await execute_procedure('insertarArtista', { nombre, ulrFoto, fecha });
        if (insert_result[0].respuesta == 0) return res.status(404).json({content: "Ya existe un artista registrado con el nombre: " + nombre });     
        
        //subo la imagen al bucket se se desea
        if ((nombreFoto != '') && (foto != '')){
          const resultS3 = await uploadS3Base64('fotos/', nombreFoto, foto, 'image');
          console.log('\n\n', resultS3.message, ', ', resultS3.url, '\n\n');      
          
          //edito al artista en la base de datos
          ulrFoto = resultS3.url;
          const id_artista = insert_result[0].respuesta;
          await  execute_procedure('editarFotoArtista', {id_artista, ulrFoto});
        }       
        
        //si todo sale correcto se devuelve un 200
        return res.status(200).json({
          content:{
            id_artista: insert_result[0].respuesta,
            nombre, 
            fecha,
            foto: ulrFoto
        } 
      })
      
    } catch (error) {
      res.status(500).json({ content: 'Internal Error al registrar artista, '+ error });
    }
}

export const editarArtista = async (req: Request, res: Response) => {
    const {id_artista, nombre, nombreFoto, foto} = req.body;
    
    try {
        //guardo al Artista en la base de datos
        const update_result = await execute_procedure('editarArtista', {id_artista, nombre});
        if (update_result[0].respuesta == 0) return res.status(404).json({content: 'Datos invalidos, puede ser que el nombre: '+nombre+' no este disponible o que el artista con el ID: ' + id_artista  +' no este registrado, porfavor registrar previamente'});
        
        //subo la imagen al bucket si se desea
        if (update_result[0].respuesta == 1 && (nombreFoto != '') && (foto != '')){
            const resultS3 = await uploadS3Base64('fotos/', nombreFoto, foto, 'image');
            console.log('\n\n', resultS3.message, ', ', resultS3.url, '\n\n');

            //edito al artista en la base de datos
            const ulrFoto = resultS3.url;
            await  execute_procedure('editarFotoArtista', {id_artista, ulrFoto});
        }

        //si todo sale correcto se devuelve un 200
        return res.status(200).json({
            content:"Se ha actualizado con exito el Artista: " + nombre
        })
    }  catch (error) {
        res.status(500).json({ content: 'Internal Error al editar Artista, '+ error });
    }
}

export const eliminarArtista = async(req: Request, res: Response) => {
    const { id_artista } = req.body;
    
    try {
        const delete_result = await execute_procedure('eliminarArtista', {id_artista});
        if (delete_result[0].respuesta == 0) return res.status(404).json({content: "No existe un artista registrado con el id: " + id_artista +', porfavor registrar previamente'});
    
        //si todo sale correcto se devuelve un 200
        return res.status(200).json({
            content:"Se ha eliminado con exito el Artista, ID: " + id_artista
        })
    } catch (error) {
        res.status(500).json({ content: 'Internal Error al eliminar artista, '+ error });
    }
}

export const obtenerArtistaxNombre = async(req: Request, res: Response) => {
    const { nombre } = req.body;
    
    try {
        const get_artistas = await execute_procedure('obtenerArtistaxNombre', {nombre});
        if (get_artistas[0].respuesta == 0) return res.status(404).json({content: "No existe un artista registrado con el nombre: " + nombre +', porfavor registrar previamente'});
    
        //si todo sale correcto se devuelve un 200
        return res.status(200).json({
            content: get_artistas
        })
    } catch (error) {
        res.status(500).json({ content: 'Internal Error al obtener artista, '+ error });
    }
}

export const obtenerArtistaxId = async(req: Request, res: Response) => {
    const { id_artista } = req.body;
    
    try {
        const get_artistas = await execute_procedure('obtenerArtistaxId', {id_artista});
        if (get_artistas[0].respuesta == 0) return res.status(404).json({content: "No existe un artista registrado con el id: " + id_artista +', porfavor registrar previamente'});
    
        //si todo sale correcto se devuelve un 200
        return res.status(200).json({
            content: get_artistas
        })
    } catch (error) {
        res.status(500).json({ content: 'Internal Error al obtener artista, '+ error });
    }
}

export const obtenerArtistas = async(req: Request, res: Response) => {
    try {
      const get_artistas = await execute_procedure('obtenerArtistas');
  
      //si todo sale correcto se devuelve un 200
      return res.status(200).json({
        content: get_artistas
      })
    } catch (error) {
      res.status(500).json({ content: 'Internal Error al obtener artistas, '+ error });
    }
}

export const healthy = (req: Request, res: Response) => {
    return res.status(200).json({ mscontentg: 'true' })
} 

File: backend_python/src/controllers/artista_controller.py
from decouple import config
from configs.mysql_config import execute_procedure
from configs.s3_config import uploadS3Base64

async def insertarArtista(data):
    try:
        urlFoto = '-'
        # Llama al procedimiento almacenado insertarArtista en Python
        # insert_result almacena el ID del usuario insertado o 0 si el usuario ya existe
        insert_result = await execute_procedure('insertarArtista', [data['nombre'], urlFoto, data['fecha']])
        if insert_result.get('respuesta') == 0:
            return {'error404': 'Ya existe un artista registrado con el nombre: ' + data['nombre']}

        if data['nombreFoto'] != '' and data['foto'] != '':
            # url_image: sintaxis por conveniencia para el nombre de la imagen
            url_image = f'fotos/{data["nombreFoto"]}'
            # Carga la imagen en S3 y obtiene la URL
            foto_url_s3 = await uploadS3Base64(data['foto'], url_image, config('aws_bucket_name'), 'image')
            id_artista = insert_result['respuesta']

            await execute_procedure('editarFotoArtista', [id_artista, foto_url_s3])

        new_artista = {
            'id_artista' : insert_result['respuesta'],
            'nombre': data['nombre'],
            'foto': urlFoto,
            'fecha': data['fecha']
        }
        return new_artista

    except Exception as e:
        # Maneja errores aquí (por ejemplo, registro de errores o devolución de un mensaje de error)
        return {'error': str(e)}
    
async def editarArtista(data):
    try:
        # Guardo al artista en la base de datos
        update_result = await execute_procedure('editarArtista', [data['id_artista'], data['nombre']])
        if update_result.get('respuesta') == 0:
            return {'error404': 'No existe un artista registrado con el ID: ' + str(data['id_artista']) +', porfavor registrar previamente'}

        # Subo la imagen al bucket si se desea
        if update_result['respuesta'] == 1 and data['nombreFoto'] != '' and data['foto'] != '':
            # url_image: sintaxis por conveniencia para el nombre de la imagen
            url_image = f'fotos/{data["nombreFoto"]}'
            # Carga la imagen en S3 y obtiene la URL
            foto_url_s3 = await uploadS3Base64(data['foto'], url_image, config('aws_bucket_name'), 'image')

            await execute_procedure('editarFotoArtista', [data['id_artista'], foto_url_s3])

        # Si todo sale correcto se devuelve un objeto artista para despues devolver un 200
        new_artista = {
            'content' : 'Se ha actualizado con exito el Artista: ' + data['nombre'],
            'nombre': data['nombre']
        }
        return new_artista

    except Exception as e:
        # Maneja errores aquí (por ejemplo, registro de errores o devolución de un mensaje de error)
        return {'error': str(e)}

async def eliminarArtista(data):
    try:
        delete_result = await execute_procedure('eliminarArtista', [data['id_artista']])
        if delete_result.get('respuesta') == 0:
            return {'error404': 'No existe un artista registrado con el ID: ' + data['id_artista'] +', porfavor registrar previamente'}

        # Si todo sale correcto se devuelve un objeto artista para despues devolver un 200
        new_artista = {
            'content' : 'Se ha eliminado con exito el Artista',
            'id': data['id_artista']
        }
        return new_artista

    except Exception as e:
        # Maneja errores aquí (por ejemplo, registro de errores o devolución de un mensaje de error)
        return {'error': str(e)}
    
async def obtenerArtistaxNombre(data):
    try:
        get_artist = await execute_procedure('obtenerArtistaxNombre', [data['nombre']])
        if get_artist.get('respuesta') == 0:
            return {'error404': 'No existe un artista registrado con el nombre: ' + data['nombre'] +', porfavor registrar previamente'}

        # Si todo sale correcto se devuelve un objeto artista para despues devolver un 200
        new_artista = {
            'content' : get_artist,
            'nombre': data['nombre']
        }
        return new_artista

    except Exception as e:
        # Maneja errores aquí (por ejemplo, registro de errores o devolución de un mensaje de error)
        return {'error': str(e)}
    
async def obtenerArtistaxId(data):
    try:
        get_artist = await execute_procedure('obtenerArtistaxId', [data['id_artista']])
        if get_artist.get('respuesta') == 0:
            return {'error404': 'No existe un artista registrado con el id: ' + data['id_artista'] +', porfavor registrar previamente'}

        # Si todo sale correcto se devuelve un objeto artista para despues devolver un 200
        new_artista = {
            'content' : get_artist,
            'id': data['id_artista']
        }
        return new_artista

    except Exception as e:
        # Maneja errores aquí (por ejemplo, registro de errores o devolución de un mensaje de error)
        return {'error': str(e)}
    
async def obtenerArtistas():
    try:
        get_artistas = await execute_procedure('obtenerArtistas')

        # Esta vacio...
        if isinstance(get_artistas, dict):
            new_artistas = {
                'content' : []
            }
            return new_artistas

        # Si todo sale correcto se devuelve un objeto artista para despues devolver un 200
        new_artistas = {
            'content' : get_artistas
        }
        return new_artistas

    except Exception as e:
        # Maneja errores aquí (por ejemplo, registro de errores o devolución de un mensaje de error)
        return {'error': str(e)}